src/Data/Array.purs

```diff
@@ -1118,13 +1146,82 @@ deleteBy eq x ys = maybe ys (\i -> unsafePartial $ fromJust (deleteAt i ys)) (fi
 -- | difference [2, 1] [2, 3] = [1]
 -- | ```
 -- |
--- | Running time: `O(n*m)`, where n is the length of the first array, and m is
+-- | Running time: `O(n+m)`, where n is the length of the first array, and m is
 -- | the length of the second.
-difference :: forall a. Eq a => Array a -> Array a -> Array a
-difference = foldr delete
+difference :: forall a. Ord a => Array a -> Array a -> Array a
+difference = differenceBy compare
 
 infix 5 difference as \\
 
+differenceBy :: forall a. (a -> a -> Ordering) -> Array a -> Array a -> Array a
+differenceBy   _ left       [] = left
+differenceBy   _ left@[]     _ = left
+differenceBy cmp left    right = map snd $ ST.run do
+  result <- STA.new
+  latestRightArrayValue <- STRef.new Nothing
+  ST.foreach indexedAndSorted \(Tuple fromLeftArray pair@(Tuple _ x')) -> do
+    maybeValueToRemove <- STRef.read latestRightArrayValue
+    case fromLeftArray, maybeValueToRemove of
+      true, Just (Tuple valueToRemove count) | cmp valueToRemove x' == EQ ->
+        -- do not add left array's element to final array; check count
+        if count == 1 then do
+          void $ STRef.write Nothing latestRightArrayValue
+        else do
+          let decrementCount = Just (Tuple valueToRemove (count - 1))
+          void $ STRef.write decrementCount latestRightArrayValue
+
+      true, _ -> void $ STA.push pair result
+
+      false, Just (Tuple valueToRemove count) | cmp valueToRemove x' == EQ -> do
+        let next = if count == 0 then Nothing
+                   else Just (Tuple valueToRemove (count - 1))
+        void $ STRef.write next latestRightArrayValue
+      false, _ -> do
+        void $ STRef.write (Just (Tuple x' 1)) latestRightArrayValue
+  _ <- STA.sortWith fst result
+  STA.unsafeFreeze result
+  where
+    -- Note: when elements are equal, right array elements
+    -- (i.e. `Tuple false (Tuple _ _)`) appear "before" left array elements
+    -- (i.e. `Tuple true (Tuple _ _)`) in the resulting array.
+    valueThenRightFirst :: Tuple Boolean (Tuple Int a) -> Tuple Boolean (Tuple Int a) -> Ordering
+    valueThenRightFirst (Tuple lb (Tuple _ lv)) (Tuple rb (Tuple _ rv)) =
+      cmp lv rv <> compare lb rb
+
+    indexedAndSorted = sortBy valueThenRightFirst $ combineIndex left right
+
+-- Internal use only
+-- Essentially...
+-- ```
+-- combineWithIndex cmp left right =
+-- let
+--   leftIndexed = mapWithIndex (\idx v -> Tuple idx v) left
+--
+--   tupleAdjusted idx v = Tuple (idx + (length left)) v
+--   rightIndexedPlus = mapWithIndex tupleAdjusted right
+--
+-- in leftIndexed <> rightIndexedPlus
+-- ```
+-- ... but without creating two intermediate arrays due to the `mapWithIndex`
+-- on both arrays.
+-- ```
+-- let t3 a b c = Tuple a (Tuple b c)
+-- combineIndex compare [0] [0, 1]
+--   == [t3 true 0 0, t3 false 1 0, t3 false 2 1]
```

Comment:
```suggestion
-- combineIndex compare [7] [9, 5]
--   == [t3 true 0 7, t3 false 1 9, t3 false 2 5]
```
Easier to follow example with values that are distinct from indices

Reply:
But then there aren't values that exist in both arrays.

Reply:
Since `combineIndex` doesn't care about duplicates, I don't think we need the example to involve duplicates.
This would be like saying that `[1,2] <> [3,4] == [1,2,3,4]` is a bad example for append because it doesn't clarify that duplicates are preserved.

Could compromise with something like `[7] [7, 5]`.

Reply:
> Since `combineIndex` doesn't care about duplicates, I don't think we need the example to involve duplicates.

:man_facepalming: You're right! I forgot that this function doesn't need to care unlike the ones I'm working on. I'll make that fix.